### PR TITLE
[vcpkg] Track hashes of cmake script helpers in consuming portfiles

### PR DESCRIPTION
**Pull request by ras0219**
This PR adds a heuristic method for determining what script helpers a given portfile is using and includes the hashes of those helpers into the port's package abi.

### CODE REVIEW

**Comment by BillyONeal** [comment]
ITK is https://github.com/microsoft/vcpkg/issues/12298

**Comment by ras0219-msft** [comment]
The arm64 errors are all due to a failure while downloading winflexbison from sourceforge.

**Comment by JackBoosY** [comment]
@ras0219-msft I think `vcpkg_find_acquire_program` also need `vcpkg_from_sourceforge`.

**Comment by JackBoosY** [comment]
rpclib:
```
CMake Error: Could not open file for write in copy operation D:/buildtrees/rpclib/src/v2.2.1-2bd00f7cbe.clean/include/rpc/version.h.tmp
CMake Error: : System Error: Permission denied
```

qt5-imageformats:
```
D:/installed/x86-windows/tools/qt5/debug/mkspecs/features/qt_plugin.prf:59: Cannot write file D:\buildtrees\qt5-imageformats\x86-windows-dbg\mkspecs\modules-inst\qt_plugin_qwbmp.pri: Cannot create parent directory
```

**Comment by Neumann-A** [comment]
>qt5-imageformats: Cannot create parent directory

If you see that one just retry. I sometimes also see it locally but just retrying seems to be enough to make it build. Probably a parallel build issue?

**Comment by PhoebeHui** [comment]
The osx failures:

libb2: looks a real issue, need a fix.
```
Makefile.am:7: error: 'pkgconfig_DATA' is used but 'pkgconfigdir' is undefined
src/Makefile.am: installing './depcomp'
parallel-tests: installing './test-driver'
autoreconf: automake failed with exit status: 1
```
spaceland: source download failed, rerun the PR might work.
ITk: confict with gdcm should be fixed by PR https://github.com/microsoft/vcpkg/pull/11208

**Comment by PhoebeHui** [comment]
Fix libb2 issue via PR#12398

File: docs/maintainers/vcpkg_common_definitions.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ VCPKG_TARGET_IMPORT_LIBRARY_PREFIX       import library prefix for target (same
 VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX       import library suffix for target (same as CMAKE_IMPORT_LIBRARY_SUFFIX)
 VCPKG_FIND_LIBRARY_PREFIXES              target dependent prefixes used for find_library calls in portfiles
 VCPKG_FIND_LIBRARY_SUFFIXES              target dependent suffixes used for find_library calls in portfiles
+VCPKG_SYSTEM_LIBRARIES                   list of libraries are provide by the toolchain and are not managed by vcpkg
```
Comment (strega-nil) [inline docs/maintainers/vcpkg_common_definitions.md:19]:
```suggestion
VCPKG_SYSTEM_LIBRARIES                   list of libraries that are provided by the toolchain and are not managed by vcpkg
```

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 2]
```diff
@@ -89,6 +90,7 @@ namespace vcpkg
 
             Lazy<std::vector<VcpkgPaths::TripletFile>> available_triplets;
             Lazy<std::vector<Toolset>> toolsets;
+            Lazy<std::map<std::string, std::string>> cmake_script_hashes;
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:93]: I don't really like this -- I would much rather have a `LazyMap` type added, so we only get the script hashes of the scripts we need. (this is not necessary for this specific PR, but it'd be good to think about for the future).

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 3]
```diff
@@ -278,6 +280,26 @@ If you wish to silence this error and use classic mode, you can:
...
+            std::map<std::string, std::string> helpers;
+            auto files = fs.get_files_non_recursive(this->scripts / fs::u8path("cmake"));
+            auto common_functions = fs::u8path("vcpkg_common_functions");
+            for (auto&& file : files)
+            {
+                auto stem = file.stem();
+                if (stem != common_functions)
+                {
+                    helpers.emplace(stem.u8string(),
+                                    Hash::get_file_hash(VCPKG_LINE_INFO, fs, file, Hash::Algorithm::Sha1));
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:296]: Is there a reason we're using Sha1 here?

**Review by strega-nil — APPROVED** [review]

File: toolsrc/src/vcpkg/vcpkgpaths.cpp (hunk 2 above)
Comment (BillyONeal) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:93]: I think we always need the whole set since we grep for the keys?

Comment (strega-nil) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:93]: hmm... right

File: toolsrc/src/vcpkg/vcpkgpaths.cpp (hunk 3 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:296]: fair enough!

**Comment by ras0219-msft** [comment]
Thanks everyone!

**Pull request merged**